### DISCUSSION TITLE
Auto-post new mastodon post and use ID from it in page?

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocs/mkdocs#3372] [Q&A]
This is a difficult thing to ask... Or at least I think it is.

I utilize Mastodon and my posts on it to create a comment system where people can reply on, which would be displayed as a commend. Here is an example:
![grafik](https://github.com/mkdocs/mkdocs/assets/11576465/55b16730-cbbb-4487-b0d6-559f859cff0e)

The main issue I currently have with this system is, that I have to first create a post on Mastodon to get the ID and then use it in my page before building and pushing it, or build and push my page, create a post to link to it and then edit my post to include the ID.
Both have their downsides and are somewhat inconvenient for me to do, so I wonder if I could automate this in some way...

What I would like to do is create a CI setup wich would do this:

- Check for any appearances of a placeholder text like `{{COMMENT_ID}}`
  - If none is found, build page and push it
- If one is found, obtain data from page (YAML Frontmatter like title) and execute a task to post a new Toot on my Mastodon account with that data and get the ID from it.
- Replace `{{COMMENT_ID}}` with the obtained ID.
- Commit and push the changes back to the repository to not have this placeholder anymore.
- Build site and push it.

This (obviously) is me being lazy, because I could just make a post and use that ID before pushing my site, but I feel like this automation could be really handy to have, especially since I could probably also include the page URL as the pages themself follow a specific structure in the final site thanks to the Material for MkDocs Blog plugin.

Do you have any recommendations on how I could do this... IF I can do this obviously.
Just FYI: I don't use nor have access to GitHub Actions here, so I can't use its features. I do have access to Woodpecker-CI tho and my current setup is [this](https://codeberg.org/Andre601/blog-source/src/branch/master/.woodpecker.yml)

**Comment by pawamoy** [comment]
Can't you make all this dynamic instead of hardcoding the comment ID in your page?

- when building a page, try to get a Mastodon comment that contains for example the page title
- if you find it, replace `{{COMMENT_ID}}` with its id
- if you don't find it, post such a comment, and replace `{{COMMENT_ID}}` with its id

It means you'll at best query Mastodon `N-pages` times at each build, and at worst `N-pages * 2` times. If that's an issue, you can cache a map of `{page.title: comment.id}` locally to drop that to 0 times, and once in while 2 times when you add a page. This can also be disabled in CI to avoid querying Mastodon.

**Reply by Andre601** [reply]
Could be risky imo, as what if there is a typo?
I'm just not sure... tho I could try this.

I just don't know how to do this....

**Reply by pawamoy** [reply]
I think you do :slightly_smiling_face: You already have code that post toots, right? Surely it's not harder to get a list of all your top-level toots (toots that have no parents), and iterate on them to find the one that contains the current page title.

And there wouldn't be a risk of typo since the toots are created automatically with the page title. Obviously if you change the title of a page, you would have to update the toot before building your docs, otherwise a new toot would be created, and previous comments wouldn't be re-attached.

And if you're still not convinced, then you can use an id of your choice instead of the page title. You just need to make sure that this id appears in the toot. For example, you could use slugs: page "My Super Blog Post" has id `super-blog-post` in the front-matter. Get all your toots, find the one that has `(post id: super-blog-post)` in its body, use its actual id (Mastodon toot id) in place of `{{COMMENT_ID}}`. Or you could use numeric ids :shrug: Or you could use the page URL. I guess you add the URL to the toots anyway?

There are many ways I guess :smile: Doing this dynamically means you don't need back and forth between publishing docs, creating toots, updating pages and publishing again.

**Reply by Andre601** [reply]
> I think you do 🙂 You already have code that post toots, right? Surely it's not harder to get a list of all your top-level toots (toots that have no parents), and iterate on them to find the one that contains the current page title.

I do not actually.

**Reply by pawamoy** [reply]
Ah, sorry, I misread your post indeed. Well, there's a Python library to interact with Mastodon: https://mastodonpy.readthedocs.io/en/stable/ :slightly_smiling_face: If you're more comfortable with other languages, you can always use them instead of Python and call a subprocess in your MkDocs hook/plugin.

**Comment by jon011235** [comment]
Hi I just build something like this for a jekyll blog: https://github.com/jon011235/mastodon-comments-autopost

I guess it would be easy enough to adapt it to something else

Happy new year!
